KEP-4974: initial proposal for deprecating v1.Endpoints (#4975)
kubernetes/enhancements: keps/sig-network/4974-deprecate-endpoints/README.md

[kubernetes/website]: https://git.k8s.io/website

## Summary

The `v1.Endpoints` API has been essentially deprecated since
EndpointSlices became GA in 1.21. Several new Service features (such
as dual-stack and topology, not to mention "services with more than
1000 endpoints") are implemented only for EndpointSlice, not for
Endpoints. Kube-proxy no longer uses Endpoints ever, for anything, and
the Gateway API conformance tests also require implementations to use
EndpointSlices.

Despite this, kube-controller-manager still does all of the work of
managing Endpoints objects for all Services, and a cluster cannot pass
the conformance test suite unless the Endpoints and EndpointSlice
Mirroring controllers are running, even though in many cases nothing
will ever look at the output of the Endpoints controller (and the
EndpointSlice Mirroring controller will never output anything).

While Kubernetes's API guarantees make it essentially impossible to
ever actually fully remove Endpoints, we should at least move toward a
world where most users run Kubernetes with the Endpoints and
EndpointSlice Mirroring controllers disabled.

## Motivation

### Goals

> - Ensure that all documentation reflects the fact that the Endpoints
>   API is deprecated and should not be used by new code, and that
>   Endpoints objects might not be created or mirrored in some clusters.

Reviewer: ```suggestion
- Ensure that end users favor EndpointSlice over Endpoints, and understand that the
  Endpoints API is deprecated.
```

Author: Hm. That feels wrong to me because we can't _directly_ cause that outcome.

Though, I guess, if we got to the end of the KEP process and users still favored Endpoints, I'd feel like the KEP had failed, so, yeah, maybe that should be the real goal...

> - Add apiserver warnings when users (other than the Endpoints
>   controller) create Endpoints objects.

Reviewer: ```suggestion
- Serve warnings (`Warning:` headers) when users, other than the Endpoints
  controller, create Endpoints objects.
```

Not sure this is a goal - that's more of a “how will we achieve“. **But** I'm not outright opposed to leaving it in.

  EndpointSlice Mirroring controllers to be running:

    - Fix tests such as `Service endpoints latency should not be very
      high` to check the latency of EndpointSlices rather than
      Endpoints, since that is what is actually important in
      real-world usage.

    - Demote tests such as `EndpointSliceMirroring should mirror a
      custom Endpoints resource through create update and delete` to
      non-conformance.

    - Split up tests such as `EndpointSlice should create Endpoints
      and EndpointSlices for Pods matching a Service` into a
      conformance test for EndpointSlice and a non-conformance test
      for Endpoints.

- Update the e2e test suite so that all remaining tests that depend on
  the outputs of the Endpoints or EndpointSlice Mirroring controllers
  are feature-tagged so they can be skipped in clusters that do not
  run those controllers, and set up a periodic e2e job running in such
  a configuration.

- Explicitly document that disabling `endpoints-controller` and/or
  `endpointslice-mirroring-controller` via kube-controller-manager's
  `--controllers` flag is a supported and conforming configuration.

>     - Figure out what to do with stale Endpoints in such a cluster.

Reviewer: These seems like a really long Goals section. How about
```markdown
- ensure that Conformance doesn't require a running controller that writes Endpoints objects

- reduce the end to end tests for Endpoints, to the minimum tests required to support Endpoints as a
  stable (but deprecated) API

- design, implement and document cleanup for stale Endpoints in clusters that still have these
  objects present
```

```
<<[UNRESOLVED kubernetes.default ]>>

- MAYBE change kube-apiserver to optionally not generate Endpoints for
  kubernetes.default, though this would require adding a
  kube-apiserver configuration option, and the benefit from removing
  just that 1 Endpoints object is small. Perhaps instead it could just
  add an annotation to the object pointing out the fact that Endpoints

>   is deprecated.

Reviewer: From experience, I can say that it would make teaching easier if there were an EndpointSlice for the `kubernetes` Service in the default namespace.

If we deprecate, we should mean it.

Author: Oh, it already creates EndpointSlices. (It has to, because kube-proxy doesn't look at Endpoints.)

The point here is that, if we consider the existence of a `default/kubernetes` Endpoints object to be an "API", then kube-apiserver's default behavior has to be to continue generating both Endpoints and EndpointSlices for it, and in that case, there's not a huge benefit to adding a command-line option to tell it to stop generating the Endpoints.

(If people think that the `default/kubernetes` Endpoints is _not_ an API, then we could just eventually stop generating it, without needing a config option.)

<<[/UNRESOLVED]>>
```

```
<<[UNRESOLVED disable-by-default ]>>

- MAYBE eventually move `endpoints-controller` and
  `endpointslice-mirroring-controller` out of the "on-by-default"
  controller list so that they become opt-in rather than opt-out.

- MAYBE eventually switch our e2e test jobs to disable those
  controllers by default, and only run a single periodic/optional job
  to run the Endpoints-specific tests.

<<[/UNRESOLVED]>>
```

### Non-Goals

- Deleting or modifying the `v1.Endpoints` API.

## Proposal

Overall, this KEP is mostly just about documentation and tests; it is
already possible to disable the Endpoints and EndpointSlice Mirroring
controllers via kube-controller-manager's `--controllers` option, and
we believe that this will have no ill effects in a vanilla Kubernetes
cluster (though, currently, it will cause the e2e tests to fail).

> 

Reviewer: If a controller is (still) making Endpoints objects, do we want to surface that somehow once the API is deprecated?

eg: `.status` on the associated Service?

Author: Yes, maybe? But...

In my first draft I suggested that it should add an annotation to each Endpoints:

```
metadata:
  annotations:
    endpoints.kubernetes.io/deprecated: >
      The Endpoints API is deprecated, and all users should use the
      EndpointSlice API instead. Endpoints objects are not guaranteed
      to be generated in all clusters.
```

but then I got to the PRR question about "Will enabling / using this feature result in increasing size or count of the existing API objects?" and was like, oh, yeah, probably we shouldn't be adding an annotation to every Endpoints when part of the motivation is to stop wasting space in etcd on objects nobody looks at...

(Yay PRR questions!)

(Then I thought about possibly hacking the annotations in on Get and stripping them back out on Update but that seemed potentially fragile...)

Reviewer: I'm OK with `kubernetes.io/deprecated` as an annotation. Even if it makes the objects larger.

```

### Endpoints Cleanup?

We do not want to leave stale `Endpoints` objects around forever if
the associated controllers are not running. (This is both a waste of
disk space and a potential source of confusion since the Endpoints
objects would quickly become out-of-date and incorrect.)

One possibility would be to just document that administrators should
delete all existing Endpoints themselves if they are going to disable
the controllers.

Another possibility would be to have kube-controller-manager do this
automatically. (Though the idea of having it automatically take action
because a controller _isn't_ enabled seems slightly magical? But
perhaps there is precedent somewhere else?)

Another possibility would be to create an `endpoints-cleanup`
controller, that could be enabled explicitly, and document that admins
should (probably) enable that controller if they disable the others.

Or perhaps the EndpointSlice controller could delete Endpoints objects
that were more than 24 hours out of date with respect to their
EndpointSlices?

In all cases, we should probably not automatically delete Endpoints

> that were not originally created by the Endpoints controller.

Reviewer: We could mark this as an unresolved point.

    Services). These should be updated to test EndpointSlices, with
    the Endpoints tests split out into separate feature-tagged tests.

  - `[It] [sig-network] Services should test the lifecycle of an
    Endpoint [Conformance]`: This just tests that the API works and
    does not test the behavior of the controllers, so it just needs to
    be feature-tagged, but not changed.

  - `[It] [sig-network] Service endpoints latency should not be very
    high [Conformance]`: This tests the latency of the Endpoints
    controller, and should be fixed to test the latency of the
    EndpointSlice controller instead, since the latency of the
    Endpoints controller has no impact on the functioning of a
    cluster.

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

- <test>: <link to test coverage>

> ### Graduation Criteria

Reviewer: We're not looking to graduate to GA here. This is a deprecation plan.

#### Alpha

> 

Reviewer: ```suggestion

- Deprecate the Endpoints API

```

- All Endpoints and EndpointSlice Mirroring tests have been demoted

>   from conformance, and we document this (e.g. via a blog post).

Reviewer: Ideally, try to cover this in the "mid point comms" blog about the upcoming release that will deprecate Endpoints.

- All official Kubernetes documentation is updated to primarily
  discuss EndpointSlices, and to mention Endpoints only as a
  deprecated API. No examples use `kubectl get endpoints` or involve
  creating an Endpoints object.

#### Beta

- (There is no requirement for any particular amount of time to have
  passed between Alpha and Beta.)

- We have a passing "no Endpoints" e2e job.

- Warnings are emitted when users create Endpoints.

- If we are going to add metrics or other "Am I using Endpoints?"
  diagnostics, then they are added and have tests.

- The documentation explains how to disable the Endpoints and
  EndpointSlice Mirroring controllers, but warns that third-party

>   components may still depend on them.

Reviewer: Bear in mind that in managed Kubernetes, people who use the cluster may have no way to disable these controllers.

- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
  - [Monitoring Requirements](#monitoring-requirements)
  - [Dependencies](#dependencies)
  - [Scalability](#scalability)
  - [Troubleshooting](#troubleshooting)
- [Implementation History](#implementation-history)
<!-- /toc -->

## Release Signoff Checklist

<!--
**ACTION REQUIRED:** In order to merge code into a release, there must be an
issue in [kubernetes/enhancements] referencing this KEP and targeting a release
milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
of the targeted release**.

For enhancements that make changes to code or processes/procedures in core
Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
Signoff checklist to be completed.

Check these off as they are completed for the Release Team to track. These
checklist items _must_ be updated for the enhancement to be released.
-->

Items marked with (R) are required *prior to targeting to a milestone / release*.

> - [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)

Reviewer 2: nit: you should mark the required ones off that you can.

Author: We really need to move this checklist out of the KEP body... For example, the first checkbox ought to be checked immediately after the KEP is first merged (with that box not checked)...

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

The `v1.Endpoints` API has been essentially (though not _actually_)
deprecated since EndpointSlices became GA in 1.21. Several new Service
features (such as dual-stack and topology, not to mention "services
with more than 1000 endpoints") are implemented only for

> EndpointSlice, not for Endpoints. Kube-proxy no longer uses Endpoints

Reviewer 3: pedantically: kube-proxy doesn't need `Endpoints` only because we do mirroring.  If we turned off mirroring, anyone who has a no-selector Service and populates `Endpoints` from some other source will no longer work, right?

Author: It's still true that kube-proxy doesn't _use_ Endpoints.

I added a new paragraph about how users don't even notice that Endpoints is semi-deprecated, because of the mirroring controller.

ever actually fully remove Endpoints, we should at least move toward a
world where most users run Kubernetes with the Endpoints and
EndpointSlice Mirroring controllers disabled.

## Motivation

### Goals

- Officially declare v1.Endpoints to be deprecated. Update
  documentation and put out appropriate communications (blog posts,
  etc) to ensure that end users are aware of this deprecation.

- Add warnings (e.g., `Warning:` headers on Endpoints create/update)
  to alert users of the fact that Endpoints is deprecated.

- Ensure that all core Kubernetes code uses EndpointSlices rather than
  Endpoints.

- Update the e2e test suite to make it possible to run it in a "no
  Endpoints controller" configuration, by rewriting some tests and
  adding feature tags to others.

- Update the conformance test suite to not require the Endpoints and
  EndpointSlice Mirroring controllers to be running, by rewriting some
  tests and demoting others from conformance.

- Explicitly document that disabling `endpoints-controller` and/or
  `endpointslice-mirroring-controller` via kube-controller-manager's
  `--controllers` flag is a supported and conforming configuration.

> - Come up with a plan to clean up stale Endpoints in clusters that

Reviewer 3: Should we add a `managed-by` label to Endpoints, so we can later detect those Endpoints objects which were writeen by endpoints controller?  That seems like something with low risk and potential reward.

Reviewer: Maybe we can use SSA field management as well or instead.

Author: We can use `ManagedFields`, but we wouldn't actually use SSA. (Nobody else is allowed to claim ownership of any of the Endpoints controller's objects, so any conflicts would always be ignored.)

> - Come up with a plan to clean up stale Endpoints in clusters that

Reviewer 2: Should this KEP not be this plan? Or are you saying that for a future stage of this KEP you will update this to have a plan?

Author: Yes, it's an UNRESOLVED point. We need a plan before the KEP is complete.

Reviewer 2: Im seeing approvals from others. https://github.com/kubernetes/enhancements/pull/4975#issuecomment-2646659917

I should’ve said I was reviewing from PRR shadow viewpoint. Just curious if you think you’ll have this approved for enhancement freeze?

  aren't running the Endpoints controller.

```
<<[UNRESOLVED kubernetes.default ]>>

> - MAYBE change kube-apiserver to optionally not generate Endpoints for

Reviewer 3: Agree

correspond to a Service with a selector.)

```
<<[UNRESOLVED endpoints-cleanup ]>>

Decide what to do here. (In the earlier stages we can just recommend
manual deletion.)

<<[/UNRESOLVED]>>
```

### Update Remaining Internal Endpoints Users

The aggregated API server and apiserver service proxying code still
make use of Endpoints. They will need to be updated to use
EndpointSlices (and, perhaps, to fall back to Endpoints for a while,
in case any users were manually creating Endpoints and not mirroring
them?).

### E2E Test Updates

There are a surprising number of e2e tests that still make use of
Endpoints, mostly because there was never any active effort to port
old tests away. These will need to be updated. See the [e2e
tests](#e2e-tests) section for more details.

### Risks and Mitigations

Obviously if a cluster contains components that read Endpoints
objects, then disabling the Endpoints controllers would break those
clusters. Given that the `v1.Endpoints` type would still exist in
these clusters, the failure mode would not be the component would fail
entirely with errors; it would just think that the Endpoints it was
looking for didn't exist ("yet"?).

> We would need to mitigate this by helping users to figure out if

Reviewer 3: We could write an empty Endpoints object with annotation indicating "this page intentionally left blank because..." ?

Reviewer: Potentially something for our recommended ValidatingAdmissionPolicies repo, too: I'm thinking that we define one accepted value for that annotation and then either warn or block on changing the annotation value away from the pattern(s) we validate.

Shouldn't block this KEP.

Author: You mean write out a blank Endpoints object for every Service? That seems like it ruins most of the benefit of the KEP.

(And writing out only a single "this page intentionally left blank" object wouldn't help components that were only looking for a specific Endpoints.)

Since we're not talking about changing the _default_ behavior of Kubernetes, I feel like we can dump this on the cluster admin. We will provide tools (metrics, events, etc) to help them figure out if anything in their cluster is using Endpoints, and then they can decide whether or not to disable the Endpoints controller. If they disable the Endpoints controller when their users were depending on it, that's their mistake, just like disabling any other feature that users were using.

Reviewer 3: Writing a single object doesn't need to watch pods, just services.  So it's net simpler.  I don't know if it would have any real impact, though?  At least any client who was waiting for Endpoints to appear would be more likely to find it and maybe  fail in a more obvious way?  And it could maybe go away after some time - the transition is what matters most to me.

More thinking aloud than concrete.

Author: That would work (a little bit) for cluster-wide controllers watching Endpoints in all namespaces, but I was worried more about apps/scripts waiting for a _specific_ Endpoints object to appear.

Reviewer 3: I meant to create an empty Endpoints object for each service with a selector.  If you were an app waiting for an Endpoints object, you would see one appear, but it would have no IPs.  My theory being that this would be more likely to result in a human's attention than just never appearing.

But I have NO evidence that it would work.

anything in their cluster depends on Endpoints.

> ## Design Details

Author: The KEP outline is sort of weird, and it's never clear what belongs in "Design Details" and what belongs in "Proposal". Also, it's not clear what, if anything is supposed to go directly under the "Design Details" header, as opposed to being part of the sub-headings.

Anyway, the current text is not an accident.

be completed all at once.

- Update all remaining internal code to use EndpointSlices rather than
  Endpoints.

- Update/reorganize e2e tests so that all tests that depend on the
  Endpoints controller are in a single test suite in
  `test/e2e/network/endpoints.go` and all tests that depend on the
  EndpointSlice mirroring controller are in a single test suite in
  `test/e2e/network/endpointslicemirroring.go`. (The latter may
  already be true.)

- Create a periodic e2e job that runs with the Endpoints and
  EndpointSlice Mirroring controllers disabled, and with the
  associated tests skipped, and confirm that it passes.

#### Deprecation - Stage 3

Stage 3 will not happen until the relevant SIGs feel that enough time
has passed since the initial deprecation announcement.

- The tests depending on the Endpoints and EndpointSlice Mirroring
  controllers are demoted from conformance, and there is additional
  communication (e.g. blog post) about this.

- The documentation is updated to explain how to disable the Endpoints
  and EndpointSlice Mirroring controllers, but warns that third-party
  components may still depend on them.

#### Deprecation - Stage 4

> Stage 4 will not happen until the relevant SIGs feel reasonably

Reviewer 2: Please call out the relevant SIGS explicitly.

then decides this was a bad idea, they can re-enable them, and even if
something had previously deleted all of the autogenerated Endpoints
objects, re-enabling the endpoints controller would regenerate them.)

###### What happens if we reenable the feature if it was previously rolled back?

N/A

###### Are there any tests for feature enablement/disablement?

N/A: The KEP does not define a new feature that can be enabled/disabled.

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

Disabling the endpoints controllers in a cluster where some
third-party components depend on them could have arbitrarily bad
effects.

###### What specific metrics should inform a rollback?

We may add metrics monitoring usage of the v1.Endpoints API, but
ideally you would have looked at those metrics _before_ disabling the
Endpoints controller.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

N/A

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

> N/A

Reviewer 2: This feature is about deprecation?

> ### Monitoring Requirements

Reviewer 2: When deprecations happen, it is very difficult for an operator to know if someone is actually using endpoints.

Are there any existing metrics to show that endpoint controller is used? 

It would be helpful from an operator to know that their is usage of a deprecated controller.

Author: (already answered in the sub-section about "are there any missing metrics")

###### How can an operator determine if the feature is in use by workloads?

> The feature would be enabled by the operator.

Reviewer 2: I don't see how this is true. We are deprecating endpoints in our code base. I did not read that the operator would trigger deprecation.

Author: Ah, the PRR section was originally written from the viewpoint of "the feature" == "disabling the controllers", and an earlier reviewer pointed out that this was wrong and I rewrote it but I guess I missed some spots.

No

###### Will enabling / using this feature result in introducing new API types?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No; the goal of this KEP is to drastically reduce the overall size of
the API database.

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No; the goal of this KEP is to drastically reduce the overall size of
the API database, and to somewhat reduce the CPU usage of
kube-controller-manager, kube-apiserver, and etcd.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

N/A

###### What are other known failure modes?

None

###### What steps should be taken if SLOs are not being met to determine the problem?

N/A

## Implementation History

- Initial proposal: 2024-11-21

> 

Reviewer 2: You are missing Drawbacks, Alternatives, and Infrastrucuture needed (Optional).

Author: There is really nothing interesting to say (and "Infrastructure needed (Optional)" is optional, and explicitly intended to be deleted if not used)

there is a stronger "API" guarantee around the existence of the
`kubernetes` Endpoints object than there is around Endpoints objects
in general.

### Endpoints Cleanup

We do not want to leave stale `Endpoints` objects around forever if
the Endpoints controllers are not running. (This is both a waste of
disk space and a potential source of confusion since the Endpoints
objects would quickly become out-of-date and incorrect.)

One possibility would be to just document that administrators should
delete all existing Endpoints themselves if they are going to disable
the controllers.

Another possibility would be to create an `endpoints-cleanup`
controller, that could be enabled explicitly, and document that admins
should (probably) enable that controller if they disable the others.

Or perhaps the EndpointSlice controller could delete Endpoints objects
that were more than 24 hours out of date with respect to their
EndpointSlices?

In all cases, we should probably not automatically delete Endpoints
that don't looke like they were originally created by the Endpoints
controller. (That is, we should not delete Endpoints unless they

> correspond to a Service with a selector.)

Reviewer 4: What if the Service was deleted in the meantime?

I think we never decided to create any annotation or use OwnerRef to reference the service, so we don't have a good mechanism to discover that programatically? Should we add it now for the purpose of doing that discovery?

Reviewer: (nit: a label seems like a better fit than an annotation)

Author: So note that in that case we _already_ end up leaving an orphan Endpoints around.

Various reviewers have suggested that we should start having the Endpoints controller label its creations (eg, https://github.com/kubernetes/enhancements/pull/4975#discussion_r1945686819).

```
<<[UNRESOLVED endpoints-cleanup ]>>

Decide what to do here. (In the earlier stages we can just recommend
manual deletion.)

<<[/UNRESOLVED]>>
```

### Update Remaining Internal Endpoints Users

The aggregated API server and apiserver service proxying code still
make use of Endpoints. They will need to be updated to use
EndpointSlices (and, perhaps, to fall back to Endpoints for a while,
in case any users were manually creating Endpoints and not mirroring

> them?).

Reviewer 4: Would endpoint-slice-mirror controller mirror them in such case?

Author: Probably, yes. FYI, there's already a PR for this (https://github.com/kubernetes/kubernetes/pull/129837) and the release note is currently:

```
When proxying to an aggregated API server, kube-apiserver now uses the
EndpointSlices of the `service` indicated by the `APIServer`, rather than
using Endpoints.

If you are using the aggregated API server feature, and you are writing out
the endpoints for it by hand (rather than letting kube-controller-manager
generate Endpoints and EndpointSlices for it automatically based on the
Service definition), then you should write out an EndpointSlice object rather
than (or in addition to) an Endpoints object.
```

Endpoints.

- The documentation becomes more bullish on the idea of disabling the
  controllers, implying that it's a reasonable default.

### Upgrade / Downgrade Strategy

The KEP does not propose any automatic change to behavior; behavior
would only be changed when the administrator chose to disable the
controllers, which could happen at any time.

### Version Skew Strategy

The only non-opt-in behavioral change is fixing the apiserver
aggregation and proxying APIs to use EndpointSlices rather than
Endpoints internally, which does not present any skew issues (since
EndpointSlices have already existed for a long time at this point).

(When the conformance criteria change to allow disabling the Endpoints
controller, this would only apply to clusters that are _fully_ at the
new version, to avoid skew issues. More specifically: it is not
conforming to disable the Endpoints controller if you still have any
apiservers that use Endpoints for apiserver aggregation and proxying.)

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

> The KEP does not define a new feature, it just proposes that we

Reviewer 4: Shouldn't we have a feature-gate for the changes we will do to stop using Endpoint in aggregation etc.?

Author: I was considering that a bugfix. The only way the behavior change could actually break someone is if they were writing out the Endpoints by hand rather than having them be generated from a Service _and_ they were explicitly setting the `skip-mirroring` annotation for some reason...

Reviewer 4: OK - I don't feel strongly enough about it to fight for FG here.

But can you please extend the answer to this question to reflect your above comment?

Author: done

document that users are allowed to disable the Endpoints and
EndpointSlice Mirroring controllers.

###### Does enabling the feature change any default behavior?

The KEP does not define a new feature that can be enabled.

(Obviously disabling the endpoints controllers changes default
behavior, but this would be because the administrator chose to do
that, not something that would happen automatically.)

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

The KEP does not define a new feature that can be enabled/disabled.

(If an administrator chooses to disable the endpoints controllers, and
then decides this was a bad idea, they can re-enable them, and even if
something had previously deleted all of the autogenerated Endpoints
objects, re-enabling the endpoints controller would regenerate them.)

###### What happens if we reenable the feature if it was previously rolled back?

> N/A

Reviewer 4: Not a blocker now, but I think that before `Stage 3`, we actually should test endpoints-controller disablement -> enablement -> disablement (again).

Can you please mention that such test will be run before stage 3 (I don't want to block you by bothering you about full description now).

Author: Sure. Is there some particular failure mode you're worried about? (Is this more about the stale Endpoints cleanup than about the Endpoints controller itself?)

the conformance test suite unless the Endpoints and EndpointSlice
Mirroring controllers are running, even though in many cases nothing
will ever look at the output of the Endpoints controller (and the
EndpointSlice Mirroring controller will never output anything).

While Kubernetes's API guarantees make it essentially impossible to
ever actually fully remove Endpoints, we should at least move toward a
world where most users run Kubernetes with the Endpoints and
EndpointSlice Mirroring controllers disabled.

## Motivation

### Goals

- Officially declare v1.Endpoints to be deprecated. Update
  documentation and put out appropriate communications (blog posts,
  etc) to ensure that end users are aware of this deprecation.

- Add warnings (e.g., `Warning:` headers on Endpoints create/update)
  to alert users of the fact that Endpoints is deprecated.

- Update the e2e test suite to make it possible to run it in a "no
  Endpoints controller" configuration, by rewriting some tests and
  adding feature tags to others.

- Update the conformance test suite to not require the Endpoints and
  EndpointSlice Mirroring controllers to be running, by rewriting some

>   tests and demoting others from conformance.

Reviewer 5: Conformance and e2e should be behavior driven so this is ok if the behavior is covered ... IIRC most of the tests use the endpoints/endpoint slices objects to synchronize the test and avoid races and flakes because of that, so a conformance test that uses these helpers should be ok to use just EndpointSlices

- Explicitly document that disabling `endpoints-controller` and/or
  `endpointslice-mirroring-controller` via kube-controller-manager's

>   `--controllers` flag is a supported and conforming configuration.

Reviewer: What guarantee did we make?

In my mind, Endpoints might go the way of ComponentStatus: still part of the OpenAPI, but it eventually stops working how you expect, per https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior

Related to this, we should make the docs clearer about that guarantee. I'm not sure I can tell what we're breaking here.

>   `--controllers` flag is a supported and conforming configuration.

Reviewer 6: > What guarantee did we make?

writing IPs to Endpoints API objects programs the network

>   `--controllers` flag is a supported and conforming configuration.

Reviewer 5: https://github.com/kubernetes/kubernetes/blob/78ad9a7b974660e5c5a1a4b4f800a9a19b4a1ddd/test/conformance/testdata/conformance.yaml#L1685-L1694

yeah, it seems is going to be complicated to justify the mirroring controller out of conformance

>   `--controllers` flag is a supported and conforming configuration.

Reviewer 6: ComponentStatus was always a read-only API, and never worked consistently across different types of cluster deployments. We marked it deprecated but have *not* changed or broken ComponentStatus.

>   `--controllers` flag is a supported and conforming configuration.

Author: I don't think writability of Endpoints is any more or less of an API guarantee than readability of Endpoints. (Meaning, if you think making `endpointslice-mirroring-controller` off-by-default is an API break then I think making `endpoints-controller` off-by-default is an API break too.)

(You might be remembering that at one point anybody could create Endpoints but only cluster-admins could create EndpointSlices, but that was only a temporary situation, and at this point you need to be cluster-admin to create either Endpoints or EndpointSlices.)

>   `--controllers` flag is a supported and conforming configuration.

Reviewer 5: I think that being able to disable endpoint controller by default and being conformance is a big win ... and we can revisit endpoint slice mirroring controller later since it seems to be more controversial and I don't think that is much of a trouble to have that controller just to mirror custom Endpoints

>   `--controllers` flag is a supported and conforming configuration.

Reviewer 6: We do need to make sure *everything* in tree is using endpoint slices and not endpoints to be able to do that, I can't remember if things like aggregated service routing or kube-apiserver service proxying switched to use endpoint slices or not

>   `--controllers` flag is a supported and conforming configuration.

Reviewer 5: > aggregated service routing

uses Endpoints https://github.com/kubernetes/kubernetes/blob/78ad9a7b974660e5c5a1a4b4f800a9a19b4a1ddd/staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go#L40-L43

> kube-apiserver service proxying

uses Endpoints

https://github.com/kubernetes/kubernetes/blob/78ad9a7b974660e5c5a1a4b4f800a9a19b4a1ddd/pkg/registry/core/service/storage/storage.go#L452-L459

>   `--controllers` flag is a supported and conforming configuration.

Reviewer 5: @danwinship for reference, list of places that need migration to EndpointSlices

```sh
$ grep -r EndpointsLister staging/ | grep -v client-go
staging/src/k8s.io/apiserver/pkg/util/proxy/proxy_test.go:              endpointLister := v1listers.NewEndpointsLister(endpointCache)
staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go:func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, namespace, id string, port int32) (*url.URL, error) {
staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go:func NewEndpointServiceResolver(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister) ServiceResolver {
staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go:  endpoints listersv1.EndpointsLister
staging/src/k8s.io/kube-aggregator/pkg/controllers/status/remote/remote_available_controller_test.go:           endpointsLister:  v1listers.NewEndpointsLister(endpointsIndexer),
staging/src/k8s.io/kube-aggregator/pkg/controllers/status/remote/remote_available_controller_test.go:                           endpointsLister:            v1listers.NewEndpointsLister(endpointsIndexer),
staging/src/k8s.io/kube-aggregator/pkg/controllers/status/remote/remote_available_controller.go:        endpointsLister v1listers.EndpointsLister
```

```sh
$ grep -r ResolveEndpoint staging/
staging/src/k8s.io/apiserver/pkg/util/proxy/proxy_test.go:              endpointURL, err := ResolveEndpoint(serviceLister, endpointLister, "one", "alfa", 443)
staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go:// ResolveEndpoint returns a URL to which one can send traffic for the specified service.
staging/src/k8s.io/apiserver/pkg/util/proxy/proxy.go:func ResolveEndpoint(services listersv1.ServiceLister, endpoints listersv1.EndpointsLister, namespace, id string, port int32) (*url.URL, error) {
staging/src/k8s.io/apiserver/pkg/util/webhook/serviceresolver.go:       ResolveEndpoint(namespace, name string, port int32) (*url.URL, error)
staging/src/k8s.io/apiserver/pkg/util/webhook/serviceresolver.go:// ResolveEndpoint constructs a service URL from a given namespace and name
staging/src/k8s.io/apiserver/pkg/util/webhook/serviceresolver.go:func (sr defaultServiceResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
staging/src/k8s.io/apiserver/pkg/util/webhook/client.go:                                u, err := cm.serviceResolver.ResolveEndpoint(cc.Service.Namespace, cc.Service.Name, port)
staging/src/k8s.io/apiserver/pkg/util/webhook/serviceresolver_test.go:                  serviceURL, err := target.ResolveEndpoint(scenario.serviceNamespace, scenario.serviceName, scenario.port)
staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/initializer/initializer_test.go:func (f *fakeServiceResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/initializer/initializer.go:   ResolveEndpoint(namespace, name string, port int32) (*url.URL, error)
staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/service_resolver.go:func (f serviceResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/service_resolver.go:func (f panickingResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
staging/src/k8s.io/apiextensions-apiserver/pkg/cmd/server/options/options.go:func (r *serviceResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler_test.go:func (dummyServiceResolverImpl) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go:  ResolveEndpoint(namespace, name string, port int32) (*url.URL, error)
staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go:func (r *aggregatorEndpointRouting) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go:  return proxy.ResolveEndpoint(r.services, r.endpoints, namespace, name, port)
staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go:func (r *aggregatorClusterRouting) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go:func (r *loopbackResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go:  return r.delegate.ResolveEndpoint(namespace, name, port)
staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:      rloc, err := r.serviceResolver.ResolveEndpoint(handlingInfo.serviceNamespace, handlingInfo.serviceName, handlingInfo.servicePort)
staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy_test.go:func (r *mockedRouter) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
staging/src/k8s.io/kube-aggregator/pkg/controllers/status/remote/remote_available_controller_test.go:func (f *fakeServiceResolver) ResolveEndpoint(namespace, name string, port int32) (*url.URL, error) {
staging/src/k8s.io/kube-aggregator/pkg/controllers/status/remote/remote_available_controller.go:        ResolveEndpoint(namespace, name string, port int32) (*url.URL, error)
staging/src/k8s.io/kube-aggregator/pkg/controllers/status/remote/remote_available_controller.go:                                discoveryURL, err := c.serviceResolver.ResolveEndpoint(apiService.Spec.Service.Namespace, apiService.Sp
```

>   `--controllers` flag is a supported and conforming configuration.

Author: >     $ grep -r ResolveEndpoint staging/

Note that `staging/src/k8s.io/kube-aggregator/pkg/apiserver/resolvers.go` is the only call to `proxy.ResolveEndpoint`; all of the other hits are unrelated methods.

>   `--controllers` flag is a supported and conforming configuration.

Reviewer 5: edited my previous comment with more accurate information

- Come up with a plan to clean up stale Endpoints in clusters that
  aren't running the Endpoints controller.

```
<<[UNRESOLVED kubernetes.default ]>>

- MAYBE change kube-apiserver to optionally not generate Endpoints for
  kubernetes.default, though this would require adding a
  kube-apiserver configuration option, and the benefit from removing
  just that 1 Endpoints object is small. Perhaps instead it could just
  add an annotation to the object pointing out the fact that Endpoints
  is deprecated.

<<[/UNRESOLVED]>>
```

```
<<[UNRESOLVED disable-by-default ]>>

> - MAYBE eventually move `endpoints-controller` and

Reviewer 6: I agree with moving towards being able to turn off Endpoints controller by default. I don't see how we'd ever turn off EndpointsMirroring controller by default without it being a hard break of behavior tied to a v1 API.

  `endpointslice-mirroring-controller` out of the "on-by-default"
  controller list so that they become opt-in rather than opt-out. If
  this was going to happen, it might make sense to create a separate
  KEP for that when we eventually do get around to it, since it will
  have additional PRR-ish concerns around upgrades, etc.

- MAYBE eventually switch our e2e test jobs to disable those
  controllers by default, and only run a single periodic/optional job
  to run the Endpoints-specific tests.

<<[/UNRESOLVED]>>
```

### Non-Goals

- Deleting or modifying the `v1.Endpoints` API.

- Removing or demoting the conformance tests that test the
  `v1.Endpoints` API independently of the Endpoints controller.

## Proposal

Overall, this KEP is mostly just about documentation and tests; it is
already possible to disable the Endpoints and EndpointSlice Mirroring
controllers via kube-controller-manager's `--controllers` option, and
we believe that this will have no ill effects in a vanilla Kubernetes
cluster (though, currently, it will cause the e2e tests to fail).

### API Server Warnings

> Whenever any Endpoints object is created or updated (except by the

Author: The mirroring controller uses the same rule as the endpoints controller does, which is that the endpoints controller owns any Endpoints object that corresponds to a Service with a selector. So the Endpoints strategy would have to keep a ServiceLister to be able to do that.

sense to annotate the EndpointSlice, since presumably anyone still
generating Endpoints objects rather than EndpointSlices is unaware of
EndpointSlices and would never look at the mirrored slice.)

The obvious problem with this idea is that part of the motivation for
this is to reduce resource usage, and adding a largish annotation to
every Endpoints object is taking one step back. (Or does etcd compress
its data in some way that would mitigate this?)

We could potentially try to have a "virtual" annotation that gets
added by the apiserver on read, and stripped back out on write, though
this might be fragile.

Having the EndpointSlice Mirroring controller modify the Endpoints
objects that it mirrors could cause also problems with controllers
that are reconciling Endpoints objects and not expecting anyone else
to modify them. (This might be problem with either "real" or "virtual"
annotations.)

#### Warnings in `kubectl get endpoints`

We could potentially have `kubectl` print warnings (to stderr) when
anyone gets or lists endpoints. There doesn't seem to be precedent for
this, but it seems like it should work.

#### Warnings via logs/metrics/Events

> We could potentially warn within the apiserver about clients Getting,

Reviewer 5: metrics labels should be bounded or you can get cardinality problems

> We could potentially warn within the apiserver about clients Getting,

Author: While it is true that this is technically unbounded, the expected cardinality is 1 (kube-controller-manager, assuming you're running the endpointslice mirroring controller) and any number larger than that indicates a problem to be fixed.

But we could have it only count the first N clients it sees...

> We could potentially warn within the apiserver about clients Getting,

Reviewer 5: yeah, good idea,  you can just do that, label as "kube-controller-manager" and then "external" or "unknown" for all the others, we do something similar in client-go requests https://github.com/kubernetes/kubernetes/blob/78ad9a7b974660e5c5a1a4b4f800a9a19b4a1ddd/staging/src/k8s.io/client-go/rest/request.go#L558-L562

> We could potentially warn within the apiserver about clients Getting,

Author: Well, not _all_ the others; the goal of this is to help the admin figure out what the offending clients are. So you'd label with the service name for the first N-1 unique clients, and then "other" for every client.

  4. Do #1, #2, or #3, but eventually (maybe a long time in the
     future) unconditionally stop generating the Endpoints object.

Whether #4 is allowed or not depends on whether we think there is an
"API"-type guarantee that a `kubernetes.default` Endpoints object
exists (or if the "API" is merely that the `kubernetes.default`
_Service_ exists and "works", without reference to _how_ it works).

```
<<[UNRESOLVED kubernetes.default ]>>

Pick an option.

<<[/UNRESOLVED]>>
```

### Endpoints Cleanup

We do not want to leave stale `Endpoints` objects around forever if
the Endpoints controllers are not running. (This is both a waste of
disk space and a potential source of confusion since the Endpoints
objects would quickly become out-of-date and incorrect.)

One possibility would be to just document that administrators should
delete all existing Endpoints themselves if they are going to disable
the controllers.

Another possibility would be to have kube-controller-manager do this
automatically. (Though the idea of having it automatically take action
because a controller _isn't_ enabled seems slightly magical? But
perhaps there is precedent somewhere else?)

> Another possibility would be to create an `endpoints-cleanup`

Reviewer 4: +1 but I would go even a step further - this controller is automatically enabled if endpoints controller is disabled (and vice versa).

I agree it's permanent, but I think it's ok.

controller, that could be enabled explicitly, and document that admins
should (probably) enable that controller if they disable the others.

Or perhaps the EndpointSlice controller could delete Endpoints objects
that were more than 24 hours out of date with respect to their
EndpointSlices?

> In all cases, we should probably not automatically delete Endpoints

Reviewer 3: Do we track that?  I don't think we have an annotation that says so?  Maybe we should add that NOW so we can use it later?

Author: "Endpoints that were not originally created by the Endpoints controller" is not _exactly_ the right concept, I guess.

We should delete all Endpoints that the Endpoints controller itself would have updated/deleted under appropriate circumstances (ie, any Endpoints that corresponds to a Service with a selector). We should not delete Endpoints that the Endpoints controller would have ignored (ie, Endpoints that do not correspond to a Service or that correspond to a selectorless Service).
